iml-api/src/graphql.rs

```diff
@@ -498,6 +487,98 @@ impl QueryRoot {
+        fqdn(description = "Pattern to search for in FQDN"),
+        tag(description = "Pattern to search for in tag"),
+        start_datetime(description = "Start of the time period of logs"),
+        end_datetime(description = "End of the time period of logs"),
+        message_class(description = "Array of log message classes"),
+        severity(description = "Upper bound of log severity"),
+    ))]
+    async fn logs(
+        context: &Context,
+        limit: Option<i32>,
+        offset: Option<i32>,
+        dir: Option<SortDir>,
+        message: Option<String>,
+        fqdn: Option<String>,
+        tag: Option<String>,
+        start_datetime: Option<String>,
+        end_datetime: Option<String>,
+        message_class: Option<Vec<MessageClass>>,
+        severity: Option<LogSeverity>,
+    ) -> juniper::FieldResult<LogResponse> {
+        tracing::info!("dir: {:?}", dir);
+        let dir = dir.unwrap_or_default();
+        tracing::info!("dir: {:?}", dir);
+
+        let start_datetime = start_datetime
+            .map(|s| s.parse::<chrono::DateTime<Utc>>())
+            .map_or(Ok(None), |r| r.map(Some))?;
+
+        let end_datetime = end_datetime
+            .map(|s| s.parse::<chrono::DateTime<Utc>>())
+            .map_or(Ok(None), |r| r.map(Some))?;
+
+        let message_class: Vec<_> = message_class
+            .filter(|v| !v.is_empty())
+            .unwrap_or_else(|| vec![MessageClass::Normal])
+            .into_iter()
+            .map(|i| i as i16)
+            .collect();
+
+        let severity = severity.unwrap_or(LogSeverity::Debug) as i16;
+
+        let results = sqlx::query_as!(
+            LogMessageRecord,
+            r#"
+                    SELECT * FROM chroma_core_logmessage t
+                    WHERE t.message LIKE $4
+                      AND t.fqdn LIKE $5
+                      AND t.tag LIKE $6
+                      AND ($7::TIMESTAMPTZ IS NULL OR t.datetime >= $7)
+                      AND ($8::TIMESTAMPTZ IS NULL OR t.datetime < $8)
+                      AND ARRAY[t.message_class] <@ $9
+                      AND t.severity <= $10
+                    ORDER BY
+                        CASE WHEN $3 = 'ASC' THEN t.datetime END ASC,
+                        CASE WHEN $3 = 'DESC' THEN t.datetime END DESC
+                    OFFSET $1 LIMIT $2"#,
+            offset.unwrap_or(0) as i64,
+            limit.map(|x| x as i64),
+            dir.deref(),
+            message.unwrap_or_else(|| "%".into()),
+            fqdn.unwrap_or_else(|| "%".into()),
+            tag.unwrap_or_else(|| "%".into()),
+            start_datetime,
+            end_datetime,
+            &message_class,
+            severity,
+        )
+        .fetch_all(&context.pg_pool)
+        .await?;
+        let xs: Vec<LogMessage> = results.into_iter().map(|x| x.into()).collect();
+
+        let total_count = sqlx::query!(
+            "SELECT chroma_core_logmessage_id_seq.last_value FROM chroma_core_logmessage_id_seq;"
```

Comment:
Technically, this is not number of entries, is it?

Reply:
It's not. It's not guaranteed that this will be 1:1 with row count (even `ON CONFLICT` statements can increment the sequence)

Reply:
I can `COUNT`, I just figured it would be slow.

Reply:
There are numerous options: https://stackoverflow.com/questions/7943233/fast-way-to-discover-the-row-count-of-a-table-in-postgresql

Reply:
COUNT is transactional, that's why it needs to full scan. But it seems it is the only way to get _exact_ number of rows.

Reply:
I've compared `COUNT` with an estimate of row count. Will continue looking, as it's clear our 600000 logs is nothing compared to number of logs in real systems. The estimate is pretty far off. The question would be however, if anyone wants to navigate to last page of pager. My guess would be no - search is much more useful.

```
chroma=# SELECT count(*) AS exact_count FROM public.chroma_core_logmessage;
 exact_count 
-------------
      623743
(1 row)

chroma=# EXPLAIN ANALYZE SELECT count(*) AS exact_count FROM public.chroma_core_logmessage;
                                                             QUERY PLAN                                                              
-------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=19458.54..19458.55 rows=1 width=8) (actual time=114.416..114.416 rows=1 loops=1)
   ->  Seq Scan on chroma_core_logmessage  (cost=0.00..17879.43 rows=631643 width=0) (actual time=0.281..74.233 rows=623776 loops=1)
 Planning time: 0.123 ms
 Execution time: 114.488 ms
(4 rows)

chroma=# SELECT reltuples::bigint AS estimate FROM   pg_class WHERE  oid = 'public.chroma_core_logmessage'::regclass;
 estimate 
----------
   577672
(1 row)

chroma=# EXPLAIN ANALYZE SELECT reltuples::bigint AS estimate FROM   pg_class WHERE  oid = 'public.chroma_core_logmessage'::regclass;
                                                         QUERY PLAN                                                          
-----------------------------------------------------------------------------------------------------------------------------
 Index Scan using pg_class_oid_index on pg_class  (cost=0.28..8.30 rows=1 width=8) (actual time=0.018..0.019 rows=1 loops=1)
   Index Cond: (oid = '31643'::oid)
 Planning time: 0.073 ms
 Execution time: 0.076 ms
(4 rows)
```

Reply:
Currently implemented version:
```
chroma=# SELECT chroma_core_logmessage_id_seq.last_value FROM chroma_core_logmessage_id_seq;
 last_value 
------------
     626418
(1 row)

chroma=# EXPLAIN ANALYZE SELECT chroma_core_logmessage_id_seq.last_value FROM chroma_core_logmessage_id_seq;
                                                      QUERY PLAN                                                       
-----------------------------------------------------------------------------------------------------------------------
 Seq Scan on chroma_core_logmessage_id_seq  (cost=0.00..1.01 rows=1 width=8) (actual time=0.022..0.023 rows=1 loops=1)
 Planning time: 0.051 ms
 Execution time: 0.050 ms
(3 rows)
```

Reply:
Counting only ids doesn't help much:
```
chroma=# EXPLAIN ANALYZE SELECT count(id) AS exact_count FROM public.chroma_core_logmessage;
                                                             QUERY PLAN                                                              
-------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=19578.01..19578.02 rows=1 width=8) (actual time=91.551..91.552 rows=1 loops=1)
   ->  Seq Scan on chroma_core_logmessage  (cost=0.00..17989.21 rows=635521 width=4) (actual time=0.023..55.729 rows=627552 loops=1)
 Planning time: 0.035 ms
 Execution time: 91.581 ms
(4 rows)

```

Reply:
Artificial limitations don't help either:
```
chroma=# EXPLAIN ANALYZE SELECT count(id) AS exact_count FROM public.chroma_core_logmessage WHERE id IS NOT NULL;
                                                             QUERY PLAN                                                              
-------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=19721.05..19721.06 rows=1 width=8) (actual time=123.174..123.174 rows=1 loops=1)
   ->  Seq Scan on chroma_core_logmessage  (cost=0.00..18120.64 rows=640164 width=4) (actual time=0.250..83.394 rows=632387 loops=1)
         Filter: (id IS NOT NULL)
 Planning time: 0.143 ms
 Execution time: 123.234 ms
(5 rows)

chroma=# EXPLAIN ANALYZE SELECT count(id) AS exact_count FROM public.chroma_core_logmessage WHERE id > 0;
                                                             QUERY PLAN                                                              
-------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=21321.46..21321.47 rows=1 width=8) (actual time=99.965..99.965 rows=1 loops=1)
   ->  Seq Scan on chroma_core_logmessage  (cost=0.00..19721.05 rows=640164 width=4) (actual time=0.004..69.472 rows=632397 loops=1)
         Filter: (id > 0)
 Planning time: 0.246 ms
 Execution time: 99.983 ms
(5 rows)
```

Reply:
I've implemented a trigger for row counting as described [here](http://www.varlena.com/GeneralBits/49.php).